### PR TITLE
[dtensor] improve allgather unpadding logic

### PR DESCRIPTION
**Pull request by wanchaol**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #103228
* __->__ #103219

This PR fixes allgather unpadding logic so that we only need to unpad
the full tensor instead of first chunking it to small tensors and unpad
individually, as we know how our padding algorithm works

### CODE REVIEW

**Review by wz337 — APPROVED** [review]
Ye. Great catch! Thanks for fixing!

File: torch/distributed/_tensor/placement_types.py [hunk 1]
```diff
@@ -251,15 +251,8 @@ def _to_replicate_tensor(
...
-            gathered_list = [
-                self._unpad_tensor(gathered_tensor, pad_size)  # type: ignore[misc]
-                if pad_size > 0
-                else gathered_tensor
-                for gathered_tensor, pad_size in zip(gathered_list, pad_sizes)
-            ]
-
-            result = torch.cat(gathered_list, dim=self.dim)
+            full_pad_size = sum(pad_sizes)
+            result = self._unpad_tensor(result, full_pad_size)
```
Comment (fduwjj) [inline torch/distributed/_tensor/placement_types.py:255]: Do we have a unit test to capture this? I mean since this is a fix so I guess some test cases might be failing?

**Review by fduwjj — APPROVED** [review]

**Review by fduwjj — APPROVED** [review]

File: torch/distributed/_tensor/placement_types.py (hunk 1 above)
Comment (wz337) [reply torch/distributed/_tensor/placement_types.py:255]: > Do we have a unit test to capture this? I mean since this is a fix so I guess some test cases might be failing?

I believe this is more of a better engineering so no test case is failing. So instead of chunking to create the tensor list and unpad each tensor, Wanchao's fix directly just unpad the all_gather result directly.

Comment (wanchaol) [reply torch/distributed/_tensor/placement_types.py:255]: Yep it's more like a improvement rather than a fix, there're already existing test capture these so I didn't add new tests. Updated the title to reflect
